examples/pytorch/rgcn/entity_sample_multi_gpu.py

```diff
@@ -1,140 +1,159 @@
-    parser.add_argument("--n-epochs", type=int, default=50,
-                        help="number of training epochs")
-    parser.add_argument("-d", "--dataset", type=str, required=True,
-                        choices=['aifb', 'mutag', 'bgs', 'am'],
-                        help="dataset to use")
-    parser.add_argument("--wd", type=float, default=5e-4,
-                        help="weight decay")
-    parser.add_argument("--fanout", type=str, default="4, 4",
-                        help="Fan-out of neighbor sampling")
-    parser.add_argument("--use-self-loop", default=False, action='store_true',
-                        help="include self feature as a special relation")
-    parser.add_argument("--batch-size", type=int, default=100,
-                        help="Mini-batch size. ")
+                           help="GPU(s) in use. Can be a list of gpu ids for multi-gpu training,"
+                                " e.g., 0,1,2,3.")
     args = parser.parse_args()
     devices = list(map(int, args.gpu.split(',')))
+    nprocs = len(devices)
+    print(f'Training with DGL built-in RGCN module with sampling using', nprocs, f'GPU(s)')
+
+    # load and preprocess dataset at master(parent) process
+    if args.dataset == 'aifb':
+        data = AIFBDataset()
+    elif args.dataset == 'mutag':
+        data = MUTAGDataset()
+    elif args.dataset == 'bgs':
+        data = BGSDataset()
+    elif args.dataset == 'am':
+        data = AMDataset()
+    else:
+        raise ValueError('Unknown dataset: {}'.format(args.dataset))
+    g = data[0]
+    num_rels = len(g.canonical_etypes)
+    category = data.predict_category
+    labels = g.nodes[category].data.pop('labels')
+    train_mask = g.nodes[category].data.pop('train_mask')
+    test_mask = g.nodes[category].data.pop('test_mask')
+    # find target category and node id
+    category_id = g.ntypes.index(category)
+    g = dgl.to_homogeneous(g)
+    node_ids = torch.arange(g.num_nodes())
+    target_idx = node_ids[g.ndata[dgl.NTYPE] == category_id]
+    # rename the fields as they can be changed by DataLoader
+    g.ndata['ntype'] = g.ndata.pop(dgl.NTYPE)
+    g.ndata['type_id'] = g.ndata.pop(dgl.NID)
+    # find the mapping (inv_target) from global node IDs to type-specific node IDs
+    inv_target = torch.empty((g.num_nodes(),), dtype=torch.int64)
+    inv_target[target_idx] = torch.arange(0, target_idx.shape[0], dtype=inv_target.dtype)
+    # avoid creating certain graph formats and train/test indexes in each sub-process to save momory
+    g.create_formats_()
+    train_idx = torch.nonzero(train_mask, as_tuple=False).squeeze()
+    test_idx = torch.nonzero(test_mask, as_tuple=False).squeeze()
+    # thread limiting to avoid resource competition
+    os.environ['OMP_NUM_THREADS'] = str(mp.cpu_count() // 2 // nprocs)
+
+    dist_data = num_rels, data.num_classes, labels, train_idx, test_idx, target_idx, inv_target
```

Comment:
Probably we should still call this `data` as it's not about distributed training.

Reply:
Right. I fixed it.